File: Difficulty: Medium/Longest Span in two Binary Arrays/longest-span-in-two-binary-arrays.cpp
class Solution {
  public:
    int longestCommonSum(vector<int> &a1, vector<int> &a2) {
        // Code here.
        int n=a1.size();
        vector<int>s1(n);
        for(int i=0;i<n;i++)
        {
            s1[i]=a1[i]-a2[i];
        }
        unordered_map<int,int>mp;
        int ans=0;
        int sum=0;
        for(int i=0;i<n;i++)
        {
            sum+=s1[i];
            if(sum==0)
            {
                ans=i+1;
                mp[sum]=i;
                continue;
            }
            if(mp.find(sum)!=mp.end())
            {
                ans=max(ans,i-mp[sum]);
            }
            else
            {
                mp[sum]=i;
            }
            
        }
        return ans;
    }
};

File: Difficulty: Medium/Longest Span in two Binary Arrays/longest-span-in-two-binary-arrays.java
class Solution {
    public int equalSumSpan(int[] a1, int[] a2) {
        // code here
        ArrayList<Integer>ls=new ArrayList<>();
        int n=a1.length;
        for(int i=0;i<n;i++)
        {
            ls.add(a1[i]-a2[i]);
        }
        HashMap<Integer,Integer>mp=new HashMap<>();
        int sum=0;
        int ans=0;
        mp.put(0,-1);
        for(int i=0;i<n;i++)
        {
           sum+=ls.get(i);
           if(mp.get(sum)!=null)
           {
               ans=Math.max(ans,i-mp.get(sum));
           }
            if(mp.get(sum)==null)
            {
                mp.put(sum,i);
            }
        }
        return ans;
        
    }
}